### PR TITLE
Custom outputs

### LINKED ISSUE #17: Add custom output algorithm
A custom output where users can specify what they want to be printed to the output stream: temperature, pressure volume, …

### PR DESCRIPTION
**Pull request by Luthaf**
This PR adds the ability for the users to specify custom output format. Fix #17.

The idea is to allow the users to tell what he is interested in, instead of adding more and more output implementations to lumol. Basically, what this does is to take a template string, and replace all parts enclosed in braces by the corresponding value, using [caldyn](https://crates.io/crates/caldyn) ([github page](https://github.com/Luthaf/caldyn)) for mathematical expression evaluation. A template string like `foo {natoms} {x[0]}` might give an output like this:

```
# custom output "foo {natoms} {x[0] + 3}"
foo 5 33.0
foo 8 23.9
foo 3 12.4
foo 1 4.56
```

Let me know what you think about this!

### CODE REVIEW

**Comment by g-bauer** [comment]
This is great! 

Could you comment on extensibility of this feature? From what I understand it should be possible to add additional arguments to parse within lumol? Do you think it is difficult to add for example computation of distances between atoms/molecules or output the center of mass positions?

As a side note, there are some missing third person 's'es in the docs 😄. Do you want me to make a review or directly correct them in this PR?

Besides that, I'd happily merge.

**Comment by Luthaf** [comment]
> Do you think it is difficult to add for example computation of distances between atoms/molecules 

Between atoms it should be as easy as `x[0] - x[5]` for getting a single distance (not counting for PBC).

> or output the center of mass positions?

This could be made available using `com[7]` syntax, and then computed on the fly. Syntax-wise, adding functions (`distance(1, 3)`) would require a bit of work on caldyn, but it should be pretty straightforward.

> As a side note, there are some missing third person 's'es in the docs 😄. Do you want me to make a review or directly correct them in this PR? 

Whatever suits you here =) I'll try to give it a look tomorrow and add some of them.

File: doc/src/input/simulations.md [hunk 1]
```diff
@@ -58,3 +59,25 @@ quantities. The available outputs are the following:
   the trajectory will be guessed from the `file` extension. Supported formats
   are documented in [chemfiles](http://chemfiles.github.io/chemfiles/)
   documentation.
+- The `Custom` output is the most powerful one, taking an user-provided
+  template string and using it to output data. The template should be given
+  as a string with the `template` key in the TOML input file.
+
+  Here are some examples of custom output templates:
+    - A constant string is reproduces as it is: `some data`;
```
Comment (g-bauer) [inline doc/src/input/simulations.md]: reproduced

File: src/core/src/out/cell.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+use std::io::prelude::*;
+use std::io;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use sys::System;
+use super::Output;
+
+/// The `CellOutput` write all the components of a cell to a file . The columns
```
Comment (g-bauer) [inline src/core/src/out/cell.rs]: writes

File: src/core/src/out/cell.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+/// in the file contains the following values: `A B C α β γ`.
```
Comment (g-bauer) [inline src/core/src/out/cell.rs]: contain

File: src/core/src/out/custom.rs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+
+    if let Ok(index) = input[(start + 1)..end].parse() {
+        return (&input[..start], index);
+    } else {
+        // invalid integer value
+        return (input, 0);
+    }
+}
+
+/// The `CustomOutput` write data into a file from an user-provided template.
```
Comment (g-bauer) [inline src/core/src/out/custom.rs]: writes

File: src/core/src/out/custom.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+///
+/// The template string can contain mathematical expressions, using some
+/// physical properties of the system. These mathematical expressions must be
+/// enclosed in braces (`{}`). Here are some examples:
+///
+/// - A constant string is reproduces as it is: `some data`;
```
Comment (g-bauer) [inline src/core/src/out/custom.rs]: reproduced

File: src/core/src/out/energy.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+use std::io::prelude::*;
+use std::io;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use super::Output;
+use utils;
+use sys::System;
+
+/// The `EnergyOutput` write the energy of the system to a text file, organized
```
Comment (g-bauer) [inline src/core/src/out/energy.rs]: writes

File: src/core/src/out/mod.rs [hunk 7]
```diff
@@ -0,0 +1,34 @@
+// Lumol, an extensible molecular simulation engine
+// Copyright (C) Lumol's contributors — BSD license
+
+//! Saving properties of a system during a simulation
+
+use sys::System;
+
+/// The `Output` trait define the interface for all the quantities outputted by
```
Comment (g-bauer) [inline src/core/src/out/mod.rs]: defines

File: src/core/src/out/mod.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+/// the simulation during the run. An Output can be a text or a binary data
+/// file, an image, a text log, …
+pub trait Output {
+    /// Function called once at the beginning of the simulation, which allow
```
Comment (g-bauer) [inline src/core/src/out/mod.rs]: allows

File: src/core/src/out/trajectory.rs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+// Copyright (C) Lumol's contributors — BSD license
+
+use std::path::Path;
+
+use super::Output;
+
+use sys::System;
+use sys::{Trajectory, TrajectoryBuilder, OpenMode, TrajectoryError};
+
+/// The `TrajectoryOutput` allow to write the trajectory of the system to a
```
Comment (g-bauer) [inline src/core/src/out/trajectory.rs]: allows

File: src/core/src/sys/system.rs [hunk 10]
```diff
@@ -122,8 +122,8 @@ impl System {
 
     /// Guess the bonds in the configuration using the chemfiles algorithm.
     ///
-    /// This function removes any existing bond, and tries to guess them using
-    /// a distance criteria. Because this function does a round trip to
+    /// This function remove any existing bond, and try to guess them using a
```
Comment (g-bauer) [inline src/core/src/sys/system.rs]: removes, tries

File: src/core/src/units.rs [hunk 11]
```diff
@@ -15,71 +15,76 @@
...
-        assert!(map.insert("Pa", 1e-40 / U_IN_KG).is_none());
-        assert!(map.insert("kPa", 1e-37 / U_IN_KG).is_none());
-        assert!(map.insert("MPa", 1e-34 / U_IN_KG).is_none());
-        assert!(map.insert("bar", 1e-35 / U_IN_KG).is_none());
-        assert!(map.insert("atm", 101325.0 * 1e-40 / U_IN_KG).is_none());
-
-        return map;
-    };
-);
+// Using a separated module because lazy_static do not support pub(crate)
```
Comment (g-bauer) [inline src/core/src/units.rs]: does

**Pull request merged**